README.md
[![developed_using](https://img.shields.io/badge/developed%20using-Jetbrains%20Goland-lightgrey)](https://www.jetbrains.com/go/)
<br/>
![GitHub](https://img.shields.io/github/license/petrjahoda/system_service) 
[![GitHub last commit](https://img.shields.io/github/last-commit/petrjahoda/system_service)](https://github.com/petrjahoda/system_service/commits/master)
[![GitHub issues](https://img.shields.io/github/issues/petrjahoda/system_service)](https://github.com/petrjahoda/system_service/issues)
<br/>
![GitHub language count](https://img.shields.io/github/languages/count/petrjahoda/system_service)
![GitHub top language](https://img.shields.io/github/languages/top/petrjahoda/system_service)
![GitHub repo size](https://img.shields.io/github/repo-size/petrjahoda/system_service)
<br/>
[![Docker Pulls](https://img.shields.io/docker/pulls/petrjahoda/system_service)](https://hub.docker.com/r/petrjahoda/system_service)
[![Docker Image Size (latest by date)](https://img.shields.io/docker/image-size/petrjahoda/system_service?sort=date)](https://hub.docker.com/r/petrjahoda/system_service/tags)
<br/>
[![developed_using](https://img.shields.io/badge/database-PostgreSQL-red)](https://www.postgresql.org) [![developed_using](https://img.shields.io/badge/runtime-Docker-red)](https://www.docker.com)

# System Service

## Description
Go service automatically creates database on its first run, check database size and disc free space. Sends email, when disc space is low.

## Installation Information
Install under docker runtime using [this dockerfile image](https://github.com/petrjahoda/system/tree/master/latest) with this command: ```docker-compose up -d```

## Implementation Information
Check the software running with this command: ```docker stats```. <br/>
System_service has to be running. No need to make a specific setup.

## Developer Information
Use software only as a [part of a system](https://github.com/petrjahoda/system) using Docker runtime.<br/>
 Do not run under linux, windows or mac on its own.
 


© 2020 Petr Jahoda


CHANGELOG.md
# Alarm Service Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

Please note, that this project, while following numbering syntax, it DOES NOT adhere
to [Semantic Versioning](http://semver.org/spec/v2.0.0.html) rules.

## Types of changes

* ```Added``` for new features.
* ```Changed``` for changes in existing functionality.
* ```Deprecated``` for soon-to-be removed features.
* ```Removed``` for now removed features.
* ```Fixed``` for any bug fixes.
* ```Security``` in case of vulnerabilities.

## [2021.3.1.12] - 2021-07-12

### Added
- Locales for breakdown statistics


## [2021.3.1.8] - 2021-07-08

### Added
- Locales for downtime statistics

## [2021.2.3.28] - 2021-06-28

### Changed
- Added userId and OrderRecordId for downtimeRecord
- Added userId for orderRecord


## [2021.2.3.23] - 2021-06-23

### Changed
- Removed AlarmRecord and DowntimeRecord creating user foreign keys

## [2021.2.3.21] - 2021-06-21

### Added
- overview locale

## [2021.2.3.15] - 2021-06-15

### Added
- web user table

### Changed
- updated for latest database version

## [2021.2.3.14] - 2021-06-14

### Added
- copyright
- updated libraries
- software to settings

## [2021.2.3.3] - 2021-06-03

### Changed
- locales for dayName

## [2021.2.2.27] - 2021-05-27

### Added
- some default data for tables

## [2021.2.2.13] - 2021-05-13

### Changed

- updated to latest go 1.16.4
- updated to latest go libraries

## [2021.2.2.3] - 2021-05-03

### Changed

- updated to latest go
- updated to latest go libraries

## [2021.2.1.27] - 2021-04-27

### Changed

- locales for menu-workplaces

## [2021.2.1.26] - 2021-04-26

### Added

- locales for power consumption

## [2021.2.1.14] - 2021-04-14

### Added

- locales for days and months, etc.

## [2021.2.1.8] - 2021-04-08

### Added

- application name to sql connection string

## [2021.2.1.7] - 2021-04-07

### Added

- creating page count table

## [2021.2.1.6] - 2021-04-06

### Changed

- updated to latest database structure

## [2021.1.3.30] - 2021-03-30

### Changed

- updating state_records table

## [2021.1.3.25] - 2021-03-25

### Added

- locales for states

## [2021.1.3.23] - 2021-03-23

### Fixed

- typos in locales

## [2021.1.3.22] - 2021-03-22

### Added

- locales for settings workplace-port details

## [2021.1.3.19] - 2021-03-19

### Added

- locales for settings workplace details

## [2021.1.3.17] - 2021-03-17

### Added

- locales for settings device details

## [2021.1.3.16] - 2021-03-16

### Added

- locales for settings user details

## [2021.1.3.15] - 2021-03-15

### Added

- locales for settings workshifts details
- locales for settings products details
- locales for settings states details

## [2021.1.3.11] - 2021-03-11

### Added

- locales for settings operation details

## [2021.1.3.9] - 2021-03-09

### Added

- locales for settings page
- locales for settings alarm details

## [2021.1.3.4] - 2021-03-04

### Changed

- capitalized locales for data page

## [2021.1.3.3] - 2021-03-03

### Added

- translation for charts selection

## [2021.1.3.2] - 2021-03-02

### Added

- translation for alarm table
- translation for breakdown table
- translation for downtime table
- translation for fault table
- translation for package table
- translation for part table
- translation for state table
- translation for user table
- translation for system stats table

## Changed

- interval for processing calculating disc size

## [2021.1.2.25] - 2021-02-25

### Added

- translation for tables
- translation for order table

## [2021.1.2.23] - 2021-02-23

### Added

- translation for data menu

## [2021.1.2.21] - 2021-02-21

### Changed

- foreign keys made from code, not from database structs

## [2021.1.2.13] - 2021-02-13

### Added

- locales for new menu

## [2021.1.2.9] - 2021-02-09

### Added

- locales for live workplace

## [2021.1.1.28] - 2021-01-28

### Added

- locales for production-downtime-poweroff

## [2021.1.1.21] - 2021-01-21

### Added

- locales for live menu

## [2020.4.3.14] - 2020-12-14

### Changed

- updated to latest go
- updated to latest libraries

## [2020.4.2.24] - 2020-11-24

### Added

- menu items for locale table

## [2020.4.2.17] - 2020-11-17

### Changed

- updated all go libraries
- updated Dockerfile
- updated create.sh
- added locale table

## [2020.4.1.26] - 2020-10-26

### Fixed

- fixed leaking goroutine bug when opening sql connections, the right way is this way

## [2020.3.2.22] - 2020-08-29

### Changed

- functions naming changed to idiomatic go (ThisFunction -> thisFunction)

## [2020.3.2.22] - 2020-08-22

### Added

- automatic go get -u all when creating docker image

## [2020.3.2.13] - 2020-08-13

### Changed

- updated to latest libraries
- updated to go 1.15 -> final program size reduced to 90%

## [2020.3.2.4] - 2020-08-04

### Fixed

- proper creating operation table

### Changed

- updated to all latest libraries
- functions and method renamed
- logging improved
- behavior improved
- code moved into multiple files

### Added

- MIT license

## [2020.3.1.30] - 2020-07-30

### Fixed

- proper closing database connections with sqlDB, err := db.DB() and defer sqlDB.Close()

### Added

- creating default operation

### Changed

- added tzdata to dockerfile

## [2020.3.1.28] - 2020-07-28

### Added

- creating timezone settings

## [2020.3.1.26] - 2020-07-26

### Changed

- for updated database structure

## [2020.3.1.22] - 2020-07-22

### Changed

- complete change to gorm v2

### Removed

- all about config files
- writing to log files

## [2020.2.2.18] - 2020-05-18

### Added

- init for actual service directory
- db.logmode(false)

## [2020.1.3.31] - 2020-03-31

### Added

- updated create.sh for uploading proper docker version automatically

## [2020.1.2.29] - 2020-02-29

### Removed

- MySQL database

### Fixed

- added sleep when creating tables and database is not available

## [2020.1.1.1] - 2020-01-01

### Added


mail.go
package main

import (
	"github.com/petrjahoda/database"
	"gopkg.in/gomail.v2"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"strconv"
	"time"
)

func sendEmail(header string) {
	logInfo("MAIN", "Sending email about: "+header)
	timer := time.Now()
	companyName, host, port, username, password, recipient := readMailSettings()
	message := gomail.NewMessage()
	message.SetHeader("From", username)
	message.SetHeader("Subject", companyName+": "+header)
	message.SetBody("text/html", "Disc space will last for just 30 days, please consider bigger disc for data.")
	message.SetHeader("To", recipient)
	d := gomail.NewDialer(host, port, username, password)
	err := d.DialAndSend(message)
	if err != nil {
		logError("MAIN", "Email not sent: "+err.Error())
		return
	}
	logInfo("MAIN", "Email sent in "+time.Since(timer).String())
}

func readMailSettings() (string, string, int, string, string, string) {
	logInfo("MAIN", "Reading mail settings")
	timer := time.Now()
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return "", "", 0, "", "", ""
	}

	var settingsHost database.Setting
	db.Where("name=?", "host").Find(&settingsHost)
	var company database.Setting
	db.Where("name=?", "company").Find(&company)
	companyName := company.Value
	host := settingsHost.Value
	var settingsPort database.Setting
	db.Where("name=?", "port").Find(&settingsPort)
	port, err := strconv.Atoi(settingsPort.Value)
	if err != nil {
		logError("MAIN", "Problem parsing port for email, using default port 587 "+err.Error())
		port = 587
	}
	var settingsUsername database.Setting
	db.Where("name=?", "username").Find(&settingsUsername)
	username := settingsUsername.Value
	var settingsPassword database.Setting
	db.Where("name=?", "password").Find(&settingsPassword)
	password := settingsPassword.Value
	var settingsEmail database.Setting
	db.Where("name=?", "email").Find(&settingsEmail)
	email := settingsEmail.Value
	logInfo("MAIN", "Mail settings read in "+time.Since(timer).String())
	return companyName, host, port, username, password, email
}


config.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

var DatabaseType string
var DatabaseIpAddress string
var DatabaseName string
var DatabasePort string
var DatabaseLogin string
var DatabasePassword string

type Config struct {
	DatabaseType string
	IpAddress    string
	DatabaseName string
	Port         string
	Login        string
	Password     string
}

func CreateConfigIfNotExists() {
	configDirectory := filepath.Join(serviceDirectory, "config")
	configFileName := "config.json"
	configFullPath := strings.Join([]string{configDirectory, configFileName}, "/")

	if _, checkPathError := os.Stat(configFullPath); checkPathError == nil {
		LogDebug("MAIN", "Config file already exists")
	} else if os.IsNotExist(checkPathError) {
		LogWarning("MAIN", "Config file does not exist, creating")
		mkdirError := os.MkdirAll(configDirectory, 0777)
		if mkdirError != nil {
			LogError("MAIN", "Unable to create directory for config file: "+mkdirError.Error())
		} else {
			LogInfo("MAIN", "Directory for config file created")
			data := Config{
				DatabaseType: "postgres",
				IpAddress:    "zapsi_database",
				DatabaseName: "zapsi3",
				Port:         "5432",
				Login:        "postgres",
				Password:     "Zps05.....",
			}
			file, _ := json.MarshalIndent(data, "", "  ")
			writingError := ioutil.WriteFile(configFullPath, file, 0666)
			LogInfo("MAIN", "Writing data to JSON file")
			if writingError != nil {
				LogError("MAIN", "Unable to write data to config file: "+writingError.Error())
			} else {
				LogInfo("MAIN", "Data written to config file")
			}
		}
	} else {
		LogError("MAIN", "Config file does not exist")
	}
}

func LoadSettingsFromConfigFile() {
	configDirectory := filepath.Join(serviceDirectory, "config")
	configFileName := "config.json"
	configFullPath := strings.Join([]string{configDirectory, configFileName}, "/")
	ConfigFile := Config{}
	for len(ConfigFile.DatabaseName) == 0 {
		readFile, err := ioutil.ReadFile(configFullPath)
		if err != nil {
			LogError("MAIN", "Problem reading config file")
			var err = os.Remove(configFullPath)
			if err != nil {
				LogError("MAIN", "Problem deleting file "+configFullPath+", "+err.Error())
				break
			}
			CreateConfigIfNotExists()
		}
		err = json.Unmarshal(readFile, &ConfigFile)
		if err != nil {
			LogError("MAIN", "Problem parsing config file, deleting config file")
			var err = os.Remove(configFullPath)
			if err != nil {
				LogError("MAIN", "Problem deleting file "+configFullPath+", "+err.Error())
				break
			}
			CreateConfigIfNotExists()
		}
	}
	DatabaseType = ConfigFile.DatabaseType
	DatabaseIpAddress = ConfigFile.IpAddress
	DatabaseName = ConfigFile.DatabaseName
	DatabasePort = ConfigFile.Port
	DatabaseLogin = ConfigFile.Login
	DatabasePassword = ConfigFile.Password
}


main.go
package main

import (
	"github.com/kardianos/service"
	"strconv"
	"time"
)

const version = "2021.2.2.13"
const serviceName = "System Service"
const serviceDescription = "Creates database and checks system data"
const config = "user=postgres password=pj79.. dbname=system host=database port=5432 sslmode=disable application_name=system_service"
const postgresConfig = "user=postgres password=pj79.. dbname=postgres host=database port=5432 sslmode=disable application_name=system_service"
const downloadInSeconds = 60

var serviceRunning = false
var processRunning = false

type program struct{}

func main() {
	logInfo("MAIN", serviceName+" ["+version+"] starting...")
	logInfo("MAIN", "© "+strconv.Itoa(time.Now().Year())+" Petr Jahoda")
	serviceConfig := &service.Config{
		Name:        serviceName,
		DisplayName: serviceName,
		Description: serviceDescription,
	}
	program := &program{}
	s, err := service.New(program, serviceConfig)
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
	err = s.Run()
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
}

func (p *program) Start(service.Service) error {
	logInfo("MAIN", serviceName+" ["+version+"] started")
	serviceRunning = true
	go p.run()
	return nil
}

func (p *program) Stop(service.Service) error {
	serviceRunning = false
	if processRunning {
		logInfo("MAIN", serviceName+" ["+version+"] stopping...")
		time.Sleep(1 * time.Second)
	}
	logInfo("MAIN", serviceName+" ["+version+"] stopped")
	return nil
}

func (p *program) run() {
	for serviceRunning {
		processRunning = true
		start := time.Now()
		logInfo("MAIN", serviceName+" ["+version+"] running")
		checkDatabase()
		databaseSizeMegaBytes := readDatabaseSize()
		lastSystemRecord := readLastSystemRecord()
		databaseGrowthInMegaBytes := databaseSizeMegaBytes - lastSystemRecord.DatabaseSizeInMegaBytes
		if time.Now().Sub(lastSystemRecord.CreatedAt).Hours() > 24 {
			logInfo("MAIN", "Database is larger than before")
			discSpaceMegaBytes := calculateFreeDiscSpace()
			estimatedDiscSpaceDays := calculateEstimatedDiscSpaceInDays(databaseGrowthInMegaBytes, discSpaceMegaBytes, time.Now().Sub(lastSystemRecord.CreatedAt).Hours())
			createNewSystemRecord(databaseSizeMegaBytes, databaseGrowthInMegaBytes, discSpaceMegaBytes, estimatedDiscSpaceDays)
			if estimatedDiscSpaceDays < 30 {
				sendEmail("Low Disc Space")
			}
		}
		sleepTime := downloadInSeconds*time.Second - time.Since(start)
		logInfo("MAIN", "Sleeping for "+sleepTime.String())
		time.Sleep(sleepTime)
		processRunning = false
	}
}
